### PR TITLE
Realization C asynSend 

### PR DESCRIPTION
**Pull request by githublaohu**

### CODE REVIEW

File: src/extern/CProducer.cpp [hunk 1]
```diff
@@ -127,6 +162,22 @@ int SendMessageSync(CProducer *producer, CMessage *msg, CSendResult *result) {
     return OK;
 }
 
+int SendMessageAsync(CProducer *producer, CMessage *msg, CSendSuccessCallback cSendSuccessCallback,CSendExceptionCallback cSendExceptionCallback){
+	if (producer == NULL || msg == NULL || cSendSuccessCallback == NULL || cSendExceptionCallback == NULL) {
+		return NULL_POINTER;
+	}
+	DefaultMQProducer *defaultMQProducer = (DefaultMQProducer *) producer;
+	MQMessage *message = (MQMessage *) msg;
+	CSendCallback* cSendCallback = new CSendCallback(cSendSuccessCallback , cSendExceptionCallback);
```
Comment (messense) [inline src/extern/CProducer.cpp]: `cSendCallback` should be deleted after `send`

Comment (githublaohu) [reply src/extern/CProducer.cpp]: See MQClientAPIImpl::sendMessageAsync and AsyncCallbackWrap::operationComplete（90 line），and AsyncCallbackWrap::onException()[ 117 line]

File: include/CProducer.h [hunk 2]
```diff
@@ -30,7 +31,7 @@ extern "C" {
 typedef struct CProducer CProducer;
 typedef int(*QueueSelectorCallback)(int size, CMessage *msg, void *arg);
```
Comment (jonnxu) [inline include/CProducer.h]: why need the parameter size, what's it function?

File: include/MQClientException.h [hunk 3]
```diff
@@ -60,6 +65,10 @@ class ROCKETMQCLIENT_API MQException : public std::exception {
 
   virtual const char* GetType() const throw() { return m_type.c_str(); }
 
+  int GetLine() const throw(){ return m_line;}
```
Comment (jonnxu) [inline include/MQClientException.h]: there will no exception for GetLine and GetFile, so need delete throw() for them

File: src/extern/CProducer.cpp [hunk 4]
```diff
@@ -67,8 +68,12 @@ class CSendCallback : public AutoDeleteSendCallBack{
 
 	}
     virtual void onException(MQException& e) {
-    	m_cSendExceptionCallback( e.what() );
-
+    	CMQException exception;
+		exception.error = e.GetError();
+		exception.line  = e.GetLine();
+		strncpy(exception.msg, e.what(), MAX_MESSAGE_ID_LENGTH - 1);
+		strncpy(exception.file, e.GetFile(), MAX_MESSAGE_ID_LENGTH - 1);
```
Comment (jonnxu) [inline src/extern/CProducer.cpp]: wrong alignment, format them

File: include/CProducer.h (hunk 2 above)
Comment (githublaohu) [reply include/CProducer.h]: QueueSelectorCallback is not asyn send  callback function

File: src/extern/CProducer.cpp [hunk 5]
```diff
@@ -62,10 +66,11 @@ class CSendCallback : public SendCallback{
 		m_cSendSuccessCallback(result);
 
 	}
-	  virtual void onException(MQException& e) {
-		  cout << "send Exception\n";
-	  }
-public:
+    virtual void onException(MQException& e) {
```
[the remark this answers is not in the document]
Comment (githublaohu) [reply src/extern/CProducer.cpp]: Already processed

File: src/extern/CProducer.cpp (hunk 4 above)
Comment (githublaohu) [reply src/extern/CProducer.cpp]: Already processed

File: include/MQClientException.h (hunk 3 above)
Comment (githublaohu) [reply include/MQClientException.h]: Already processed

File: include/CMQException.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#include "CCommon.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct _CMQException_{
+	int error;
+	int line;
+	char file[MAX_MESSAGE_ID_LENGTH];
```
[the remark this answers is not in the document]
Comment (githublaohu) [reply include/CMQException.h]: Already processed

File: src/extern/CProducer.cpp [hunk 7]
```diff
@@ -45,6 +50,36 @@ class SelectMessageQueue : public MessageQueueSelector {
...
 
+class CSendCallback : public AutoDeleteSendCallBack{
+public:
+	CSendCallback(CSendSuccessCallback cSendSuccessCallback,CSendExceptionCallback cSendExceptionCallback){
+		m_cSendSuccessCallback = cSendSuccessCallback;
+		m_cSendExceptionCallback = cSendExceptionCallback;
+	}
+	virtual ~CSendCallback(){}
+	virtual void onSuccess(SendResult& sendResult) {
+		cout << "send onSuccess\n";
```
[the remark this answers is not in the document]
Comment (githublaohu) [reply src/extern/CProducer.cpp]: thanks , Already processed

File: include/MQClientException.h [hunk 8]
```diff
@@ -21,12 +21,17 @@
 #include <ostream>
 #include <sstream>
 #include <string>
+
+#include <string.h>
 #include "RocketMQClient.h"
+#include "CCommon.h"
```
[the remark this answers is not in the document]
Comment (githublaohu) [reply include/MQClientException.h]: Already processed

File: include/CMQException.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+#endif
+
+#define  MAX_EXEPTION_CHAR_LENGTH 256
+
+typedef struct _CMQException_{
+	int error;
+	int line;
+	char file[MAX_EXEPTION_CHAR_LENGTH];
+	char msg[MAX_EXEPTION_CHAR_LENGTH];
+	char type[MAX_EXEPTION_CHAR_LENGTH];
```
Comment (ShannonDing) [inline include/CMQException.h]: it is better to define the MACRO separately, and the length of msg maybe bigger than 256, IMO, maybe 512 is better.

File: example/CAsyncProducer.c [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    int i = 0;
+    char DestMsg[256];
+    CMessage *msg = CreateMessage("T_TestTopic");
+    SetMessageTags(msg, "Test_Tag");
+    SetMessageKeys(msg, "Test_Keys");
+    CSendResult result;
+    for (i = 0; i < 10; i++) {
+        printf("send one message : %d\n", i);
+        memset(DestMsg, 0, sizeof(DestMsg));
+        snprintf(DestMsg, 255, "New message body: index %d", i);
```
[the remark this answers is not in the document]
Comment (jonnxu) [reply example/CAsyncProducer.c]: DestMsg chang to body?  variable need keep the same style, that is first word for variable with lowercase

**Pull request merged**